pkg/liveshare/session_test.go

```diff
@@ -400,26 +400,42 @@ func TestSessionHeartbeat(t *testing.T) {
 }
 
 func TestRebuild(t *testing.T) {
-	requestCount := 0
-	getSharedServers := func(conn *jsonrpc2.Conn, req *jsonrpc2.Request) (interface{}, error) {
-		requestCount++
-		return true, nil
-	}
-	testServer, session, err := makeMockSession(
-		livesharetest.WithService("IEnvironmentConfigurationService.rebuildContainer", getSharedServers),
-	)
-	if err != nil {
-		t.Fatalf("creating mock session: %v", err)
+	tests := []struct {
+		fullRebuild bool
+		rpcService  string
+	}{
+		{
+			fullRebuild: false,
+			rpcService:  "IEnvironmentConfigurationService.incrementalRebuildContainer",
+		},
+		{
+			fullRebuild: true,
+			rpcService:  "IEnvironmentConfigurationService.rebuildContainer",
+		},
 	}
-	defer testServer.Close()
 
-	err = session.RebuildContainer(context.Background())
-	if err != nil {
-		t.Fatalf("rebuilding codespace via mock session: %v", err)
-	}
+	for _, tt := range tests {
+		requestCount := 0
+		getSharedServers := func(conn *jsonrpc2.Conn, req *jsonrpc2.Request) (interface{}, error) {
+			requestCount++
+			return true, nil
+		}
+		testServer, session, err := makeMockSession(
+			livesharetest.WithService(tt.rpcService, getSharedServers),
+		)
+		if err != nil {
+			t.Fatalf("creating mock session: %v", err)
```

Comment:
@cmbrose Should these be fatal when I'm using a `tt` array?

Reply:
If you use `Fatal` it will stop all remaining tests in `tt` which is annoying because it can lead to the case where the test reports 1 failure, you fix it, and then the next item fails. Generally I only use `Fatal` when it's needed to avoid compiler errors about potential null refs/etc.

Just a suggestion, using `t.Log` with `t.Error` will be cleaner to read error logs about what tests are failing when using `tt` pattern

Reply:
Thanks for the explanation